R/describe-feature.R

```diff
@@ -0,0 +1,82 @@
+# Copyright 2019 Province of British Columbia
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+#' Describe a WFS feature
+#'
+#' Describe the columns from a WFS feature. The column name, whether a column can be
+#' separated from the record in WFS (nillable) and the tpye of column are returned.
+#' This can be a helpful tool to examine a layer before issuing a query with `bcdc_get_geodata`
+#'
+#' @inheritParams bcdc_get_geodata
+#' @export
+#'
+#' @examples
+#'  bcdc_describe_feature("bc-airports")
+
+bcdc_describe_feature <- function(x = NULL){
+  obj <- bcdc_get_record(x)
+  if (!"wms" %in% vapply(obj$resources, `[[`, "format", FUN.VALUE = character(1))) {
+    stop("No wms/wfs resource available for this dataset.",
+         call. = FALSE
+    )
+  }
+
+  ## Parameters for the API call
+  query_list <- list(
+    SERVICE = "WFS",
+    VERSION = "2.0.0",
+    REQUEST = "DescribeFeatureType",
+    #outputFormat = "application/json",
+    typeNames = obj$layer_name
+  )
+
+
+  ## GET and parse data to sf object
+  cli <- bcdc_http_client(url = "https://openmaps.gov.bc.ca/geo/pub/wfs")
+
+  cc <- cli$get(query = query_list)
+  status_failed <- cc$status_code >= 300
+
+  xml_res <- xml2::read_xml(cc$parse("UTF-8"))
+  xml_res <- xml2::xml_find_all(xml_res, "//xsd:sequence")
+  xml_res <- xml2::xml_find_all(xml_res, ".//xsd:element")
+  xml_res <-  purrr::map(xml_res, xml2::xml_attrs)
+  xml_df <- purrr::map_df(xml_res, ~as.list(.))
+
+  ## This is an ugly way of doing this
+  ## Manually add id and turn into a row
+  id_row <- dplyr::tibble(name = "id",
+                          nillable = FALSE,
+                          type = "xsd:string")
+
+  ## Extracting geometry column and turn into a row
+  geom_row <- xml_df[xml_df$type == "gml:GeometryPropertyType",]
```

Comment:
This is fragile - for example, `bcdc_query_geodata("bc-major-cities-points-1-2-000-000-digital-baseline-mapping")` bonks because the geometry column is of type: `gml:PointPropertyType`

Reply:
This now break the print method (in a useful way):
```
bcdc_query_geodata("bc-major-cities-points-1-2-000-000-digital-baseline-mapping") %>% 
 select(geometry)
# A B.C. Data Catalogue Record: NA records and 11 columns
# Columns:
# A tibble: 11 x 4
   col_name   selectable remote_col_type          local_col_type
   <chr>      <lgl>      <chr>                    <chr>         
 1 id         FALSE      xsd:string               character     
 2 FCODE      FALSE      xsd:string               character     
 3 BCMJ_TAG   FALSE      xsd:decimal              numeric       
 4 NAME       FALSE      xsd:string               character     
 5 CITY_TYPE  FALSE      xsd:string               character     
 6 LONG_TYPE  FALSE      xsd:string               character     
 7 POP_2000   FALSE      xsd:decimal              numeric       
 8 POP_SOURCE FALSE      xsd:string               character     
 9 OBJECTID   FALSE      xsd:decimal              numeric       
10 geometry   TRUE       gml:PointPropertyType    sfc geometry  
11 geometry   TRUE       gml:GeometryPropertyType sfc geometry  
```
It is because I didn't account for additional gml types. I'll create an issue rather than hold up this PR.

Reply:
This is the bit that causes the repetition:

https://github.com/bcgov/bcdata/blob/ffe8f733234fc0fa87af012c38d98c1cc5c149ec/R/utils-classes.R#L30-L38

 though I'm not quite sure what it does since this (`feature_spec <- bcdc_describe_feature(id)`) runs right beforehand

Reply:
Resolving this conversation in favour of the issue #50 you opened